### PR TITLE
fix #18627(Program segfaults with ARC when using openArray[string])

### LINKED ISSUE #18627: Program segfaults with ARC when using openArray[string]
Program segfaults with ARC when using openArray[string]. Compiling with `nim r -d:debug --gc:arc main.nim`. 
When using `int` instead of `string` it works. If compiling with `--gc:refc` instead of `--gc:arc` it works.

### Example
```nim
proc setPosition(params: openArray[string]) =
    for i in params.toOpenArray(0, params.len - 1):
        echo i

proc uciLoop*() =
    let params = @[readLine(stdin)]
    setPosition(params)

uciLoop()
```

### Current Output
Input is `hi`and then enter.
```
hi
Traceback (most recent call last)
/home/tsoj/Dokumente/Projects/main/main.nim(9) main
/home/tsoj/Dokumente/Projects/main/main.nim(7) uciLoop
/home/tsoj/Dokumente/Projects/main/main.nim(7) setPosition
/home/tsoj/.choosenim/toolchains/nim-#devel/lib/system/alloc.nim(973) dealloc
/home/tsoj/.choosenim/toolchains/nim-#devel/lib/system/alloc.nim(872) rawDealloc
/home/tsoj/.choosenim/toolchains/nim-#devel/lib/system/avltree.nim(74) del
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: '/home/tsoj/.cache/nim/main_d/main_7E5469BB05FCB4740D4D1F214355B678BA58330E '
```

### Expected Output
```
hi
```
### Additional Information

Possibly related is this example:
```nim
proc uciLoop*() =
    let params = @[readLine(stdin)]    
    for i in params.toOpenArray(0, params.len - 1):
        echo i
uciLoop()
```
Which compiles and runs as expected with `nim r -d:debug --gc:refc Nalwald.nim` but doesn't compile with `nim r -d:debug --gc:arc Nalwald.nim`:
```
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/nim.cfg' [Conf]
Hint: used config file '/home/tsoj/.choosenim/toolchains/nim-#devel/config/config.nims' [Conf]
......................................................
/home/tsoj/Dokumente/Projects/main/main.nim(3, 20) Error: internal error: '=destroy' operator not found for type openArray[string]
```
```
$ nim -v
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-07-27
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 22776c410a916b29c05b798523c94c834c7371ad
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by ringabout**
fix #18627

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -855,7 +855,12 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode): PNode =
         result[0] = p(n[0], c, s, normal)
       if canRaise(n[0]): s.needsTry = true
       if mode == normal:
-        result = ensureDestruction(result, n, c, s)
+        if n[0].kind == nkSym and n[0].sym.magic == mSlice:
+          # toOpenArray borrows data form owner, so we cannot
+          # destroy the result.
+          discard
+        else:
+          result = ensureDestruction(result, n, c, s)
```
Comment (Araq) [inline compiler/injectdestructors.nim]: Why does `ensureDestruction` produce a call to the destructor? Why do openArrays even have destructors?

**Comment by Araq** [comment]
Very good bugfix but the root cause lies deeper, I think.

File: compiler/injectdestructors.nim (hunk 1 above)
Comment (ringabout) [reply compiler/injectdestructors.nim]: Yep, thanks! Now I see a deeper cause.

File: compiler/ast.nim [hunk 2]
```diff
@@ -1591,7 +1591,7 @@ proc propagateToOwner*(owner, elem: PType; propagateHasAsgn = true) =
   if mask != {} and propagateHasAsgn:
     let o2 = owner.skipTypes({tyGenericInst, tyAlias, tySink})
     if o2.kind in {tyTuple, tyObject, tyArray,
-                   tySequence, tySet, tyDistinct, tyOpenArray, tyVarargs}:
+                   tySequence, tySet, tyDistinct, tyVarargs}:
```
Comment (ringabout) [inline compiler/ast.nim]:
```nim
  let mask = elem.flags * {tfHasAsgn, tfHasOwned}
  if mask != {} and propagateHasAsgn:
    let o2 = owner.skipTypes({tyGenericInst, tyAlias, tySink})
    if o2.kind in {tyTuple, tyObject, tyArray,
                   tySequence, tySet, tyDistinct, tyVarargs}:
      o2.flags.incl mask
      owner.flags.incl mask
```
Before if `openArray` contains seq/string etc, its flags will include `tfHasAsgn` too.

And here is the definition of `hasDestructor`.
```nim
template hasDestructor*(t: PType): bool = {tfHasAsgn, tfHasOwned} * t.flags != {}
```

Comment (xflywind) [reply compiler/ast.nim:1594]: As a bonus, I find another bug
```
proc setPosition(params: varargs[string]) =
    for i in params.toOpenArray(0, params.len - 1):
        echo i

proc uciLoop*() =
    let params = @[readLine(stdin)]
    setPosition(params)

uciLoop()
```
gives
```
Error: internal error: '=destroy' operator not found for type openArray[string]
```

Comment (Araq) [reply compiler/ast.nim]: Also remove the `tyVarargs` case, varargs's runtime behavior is identical to openArray.

Comment (ringabout) [reply compiler/ast.nim]: yeah, I guess it is less harmful for `tyVarargs` becuase it cannot be used as return types.

**Pull request merged**